jplag/src/main/java/de/jplag/GreedyStringTiling.java

```diff
@@ -4,100 +4,56 @@
-        if (hashLength < 1) {
-            hashLength = 1;
-        }
-        hashLength = (hashLength < 26 ? hashLength : 25);
-
-        if (tokenList.size() < hashLength) {
-            return;
-        }
-
-        int modulo = ((1 << 6) - 1);   // Modulo 64!
-
-        int loops = tokenList.size() - hashLength;
-        tokenList.tokenHashes = (makeTable ? new TokenHashMap(3 * loops) : null);
-        int hash = 0;
-        int hashedLength = 0;
-        for (int i = 0; i < hashLength; i++) {
-            hash = (2 * hash) + (tokenList.getToken(i).type & modulo);
-            hashedLength++;
-            if (tokenList.getToken(i).isMarked()) {
-                hashedLength = 0;
-            }
-        }
-        int factor = (hashLength != 1 ? (2 << (hashLength - 2)) : 1);
-
-        if (makeTable) {
-            for (int i = 0; i < loops; i++) {
-                if (hashedLength >= hashLength) {
-                    tokenList.getToken(i).setHash(hash);
-                    tokenList.tokenHashes.put(hash, i);   // add into hashtable
-                } else {
-                    tokenList.getToken(i).setHash(-1);
-                }
-                hash -= factor * (tokenList.getToken(i).type & modulo);
-                hash = (2 * hash) + (tokenList.getToken(i + hashLength).type & modulo);
-                if (tokenList.getToken(i + hashLength).isMarked()) {
-                    hashedLength = 0;
-                } else {
-                    hashedLength++;
-                }
-            }
-        } else {
-            for (int i = 0; i < loops; i++) {
-                tokenList.getToken(i).setHash((hashedLength >= hashLength) ? hash : -1);
-                hash -= factor * (tokenList.getToken(i).type & modulo);
-                hash = (2 * hash) + (tokenList.getToken(i + hashLength).type & modulo);
-                if (tokenList.getToken(i + hashLength).isMarked()) {
-                    hashedLength = 0;
-                } else {
-                    hashedLength++;
-                }
-            }
-        }
-        tokenList.hashLength = hashLength;
+        this.minimumMatchLength = options.minimumTokenMatch();
     }
 
     public final JPlagComparison compare(Submission firstSubmission, Submission secondSubmission) {
-        return swapAndCompare(firstSubmission, secondSubmission, false);
+        return swapAndCompare(firstSubmission, secondSubmission);
     }
 
-    public final JPlagComparison compareWithBaseCode(Submission firstSubmission, Submission secondSubmission) {
-        return swapAndCompare(firstSubmission, secondSubmission, true);
+    public final JPlagComparison compareWithBaseCode(Submission submission, Submission baseCodeSubmission) {
+        JPlagComparison comparison = swapAndCompare(submission, baseCodeSubmission);
+        // Remove the hashLookupTable as the isBaseCode tagging for the tokens changed which will affect the computed hashes.
```

Comment:
I don't really understand the sentence...

Reply:
```suggestion
        // Remove the lookup table for the current submission to trigger a regeneration as hashes will change due to the new baseCodeMarking.
```

Better like that? What I want to say is that we remove the computed hash lookup table for the submission we just computed the base code markings for, in order to trigger a regeneration of that table which considers the base code markings. As then said, this is done to increase performance, as subsequences with baseCode-marked tokens receive the `NO_VALUE` hash to ignore them during selection of possibly matching subsequences. Obviously, before computing the base code markings, the hash lookup table can't consider base code tokens and thus the hash lookup table computed during the comparison with the base code itself is not optimal.

Reply:
Ah, now I get it! Way better, thanks! 😄